fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -912,76 +979,100 @@ public LuceneRecordCursor getOldestNDocuments(@Nonnull final LucenePartitionInfo
+                    // shortcut delete docs from current partition
+                    // (we do this, instead of calling LuceneIndexMaintainer.update() in order to avoid a chicken-and-egg
+                    // situation with the partition metadata keys.
+                    records.forEach(r -> {
+                        try {
+                            indexMaintainer.deleteDocument(groupingKey, partitionInfo.getId(), r.getPrimaryKey());
+                        } catch (IOException e) {
+                            throw new RecordCoreException(e);
+                        }
+                    });
+
+                    // reset partition info
+                    state.context.ensureActive().clear(partitionMetadataKeyFromPartitioningValue(groupingKey, getPartitionKey(partitionInfo)));
+                    if (!repartitioningContext.emptyingPartition) {
+                        // update current partition's meta
+                        LucenePartitionInfoProto.LucenePartitionInfo.Builder builder = partitionInfo.toBuilder()
+                                .setCount(partitionInfo.getCount() - records.size());
+                        if (removingOldest) {
+                            builder.setFrom(ByteString.copyFrom(Objects.requireNonNull(newBoundaryPartitionKey).pack()));
+                        } else {
+                            builder.setTo(ByteString.copyFrom(Objects.requireNonNull(newBoundaryPartitionKey).pack()));
+                        }
+                        savePartitionMetadata(groupingKey, builder);
                     }
 
-                    Iterator<? extends FDBIndexableRecord<Message>> recordIterator = records.iterator();
-                    return AsyncUtil.whileTrue(() -> indexMaintainer.update(null, recordIterator.next())
-                            .thenApply(ignored -> recordIterator.hasNext()), state.context.getExecutor());
-                }).thenApply(ignored -> records.size());
-            }
-            return CompletableFuture.completedFuture(0);
-        });
+                    if (LOGGER.isDebugEnabled()) {
+                        LOGGER.debug(repartitionLogMessage("Repartitioned records", groupingKey, records.size(), partitionInfo));
+                    }
+
+                    // value of the "destination" partition's `from` value
+                    final Tuple overflowPartitioningKey = toPartitionKey(records.get(0));
+                    final CompletableFuture<LucenePartitionInfoProto.LucenePartitionInfo> futureDestination;
+                    if (removingOldest) {
+                        futureDestination = findPartitionInfo(groupingKey, overflowPartitioningKey);
+                    } else {
+                        futureDestination = getNextNewerPartitionInfo(state.context, groupingKey, overflowPartitioningKey, state.indexSubspace);
+                    }
```

Comment:
I think the argument was that this would create two much complexity, and have to take care to handle the partition deletion that could happen above, but the only partition deletion that could happen is the current partition, so newer and older should always be there. It's probably not a big deal, but it feels like this logic makes it harder to have confidence based solely on unit tests of the `LuceneRepartitionPlanner`.